src/modules/musig/keyagg_impl.h

```diff
@@ -0,0 +1,293 @@
+
+#ifndef SECP256K1_MODULE_MUSIG_KEYAGG_IMPL_H
+#define SECP256K1_MODULE_MUSIG_KEYAGG_IMPL_H
+
+#include <string.h>
+
+#include "keyagg.h"
+#include "../../eckey.h"
+#include "../../ecmult.h"
+#include "../../field.h"
+#include "../../group.h"
+#include "../../hash.h"
+#include "../../util.h"
+
+static const unsigned char secp256k1_musig_keyagg_cache_magic[4] = { 0xf4, 0xad, 0xbb, 0xdf };
+
+/* A keyagg cache consists of
+ * - 4 byte magic set during initialization to allow detecting an uninitialized
+ *   object.
+ * - 64 byte aggregate (and potentially tweaked) public key
+ * - 64 byte "second" public key (set to the point at infinity if not present)
+ * - 32 byte hash of all public keys
+ * - 1 byte the parity of the internal key (if tweaked, otherwise 0)
+ * - 32 byte tweak
+ */
+/* Requires that cache_i->pk is not infinity. */
+static void secp256k1_keyagg_cache_save(secp256k1_musig_keyagg_cache *cache, secp256k1_keyagg_cache_internal *cache_i) {
+    unsigned char *ptr = cache->data;
+    memcpy(ptr, secp256k1_musig_keyagg_cache_magic, 4);
+    ptr += 4;
+    secp256k1_ge_to_bytes(ptr, &cache_i->pk);
+    ptr += 64;
+    secp256k1_ge_to_bytes_ext(ptr, &cache_i->second_pk);
+    ptr += 64;
+    memcpy(ptr, cache_i->pk_hash, 32);
+    ptr += 32;
+    *ptr = cache_i->parity_acc;
+    ptr += 1;
+    secp256k1_scalar_get_b32(ptr, &cache_i->tweak);
+}
+
+static int secp256k1_keyagg_cache_load(const secp256k1_context* ctx, secp256k1_keyagg_cache_internal *cache_i, const secp256k1_musig_keyagg_cache *cache) {
+    const unsigned char *ptr = cache->data;
+    ARG_CHECK(secp256k1_memcmp_var(ptr, secp256k1_musig_keyagg_cache_magic, 4) == 0);
+    ptr += 4;
+    secp256k1_ge_from_bytes(&cache_i->pk, ptr);
+    ptr += 64;
+    secp256k1_ge_from_bytes_ext(&cache_i->second_pk, ptr);
+    ptr += 64;
+    memcpy(cache_i->pk_hash, ptr, 32);
+    ptr += 32;
+    cache_i->parity_acc = *ptr & 1;
+    ptr += 1;
+    secp256k1_scalar_set_b32(&cache_i->tweak, ptr, NULL);
+    return 1;
+}
+
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("KeyAgg list")||SHA256("KeyAgg list"). */
+static void secp256k1_musig_keyagglist_sha256(secp256k1_sha256 *sha) {
+    secp256k1_sha256_initialize(sha);
+
+    sha->s[0] = 0xb399d5e0ul;
+    sha->s[1] = 0xc8fff302ul;
+    sha->s[2] = 0x6badac71ul;
+    sha->s[3] = 0x07c5b7f1ul;
+    sha->s[4] = 0x9701e2eful;
+    sha->s[5] = 0x2a72ecf8ul;
+    sha->s[6] = 0x201a4c7bul;
+    sha->s[7] = 0xab148a38ul;
+    sha->bytes = 64;
+}
+
+/* Computes pk_hash = tagged_hash(pk[0], ..., pk[np-1]) */
+static int secp256k1_musig_compute_pk_hash(const secp256k1_context *ctx, unsigned char *pk_hash, const secp256k1_pubkey * const* pk, size_t np) {
+    secp256k1_sha256 sha;
+    size_t i;
+
+    secp256k1_musig_keyagglist_sha256(&sha);
+    for (i = 0; i < np; i++) {
+        unsigned char ser[33];
+        size_t ser_len = sizeof(ser);
```

Comment:
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Could move this outside the loop to make this slightly more performant and, imo, more readable. Not sure the performance matters but could make a difference if computing `pk_hash` for a large number of keys?

Reply:
I think that would be less readable than declaring the variables only in the block where they are needed. The compiler will almost certainly optimize the current version to not cause any overhead.